### PR TITLE
FIX: ISXB-925 Fixed an issue where changing InputSettings instance fo…

### PR DESCRIPTION
**Pull request by ekcoh**
…r the Input System wouldn't affect feature flags.

### Description

The optimization flags within InputSettings were stored as static fields for performance reasons. This makes little sense from API perspective where these are set and evaluated using a non-static API. This also leads to switching InputSettings instance having no impact on current settings. As long as these are exposed via this API they should not be static. However, I added extraction of the settings into InputManager which is represented by a static and used direct access from where its used so expecting little if any impact.

Adressing:
- Feature flags set and checked via InputSettings non-static interface but backed by static variables.

### Changes made

- Changed flags to be extracted in Input Manager instead when settings change.
- Removed custom handling in InputSettings which allow IsFeatureEnabled to work.

Note that this fix also includes https://github.com/Unity-Technologies/InputSystem/pull/1955 since the test case would otherwise fail with debug error log on non Windows platforms.

### Testing

Added 1 test case for scenario `Settings_ShouldStoreSettingsAndFeatureFlags(string featureName)`. I didn't add a new test category at this point since none of the existing fits well so I left the test case uncategorised.

### Risk

Potential risk is features not being enabled correctly for some corner-case scenarios.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.	
    - Jira port for the next release set as "Resolved".
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

### CODE REVIEW

**Review by graham-huws — APPROVED** [review]

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 1]
```diff
@@ -2116,6 +2116,12 @@ internal struct AvailableDevice
         internal IInputRuntime m_Runtime;
         internal InputMetrics m_Metrics;
         internal InputSettings m_Settings;
+
+        // Extract as booleans (from m_Settings) because feature check is in the hot path
+        internal bool m_OptimizedControlsFeatureEnabled;
+        internal bool m_ReadValueCachingFeatureEnabled;
+        internal bool m_ParanoidReadValueCachingChecksEnabled;
```
Comment (timkeo) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Could you please change these to private fields and add an internal read-only accessor property?

When refactoring for CoreCLR support, internal fields were very frustrating because I had to check the entire code base for any usages.

**Review by timkeo — APPROVED** [review]
From [my FEPM refactoring worksheet,](https://docs.google.com/spreadsheets/d/1FHIpLELBkFgMURChojRbHsPA1jKEVlT0EaFbvwQLYUQ/edit?gid=1391290404#gid=1391290404) I didn't touch these static fields, since determined it wasn't necessary for CoreCLR support. However, there may still be conflicts.

I recommend checking these changes against [my FEPM refactor branch](https://github.com/Unity-Technologies/InputSystem/pull/1915) to make sure isn't anything too complicated, otherwise porting this to Neutron will be a pain.

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs (hunk 1 above)
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: This have been adressed now in https://github.com/Unity-Technologies/InputSystem/pull/1954/commits/c93e565d0b9260c0911f2af882871dd8551fe80a

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Regarding FWPM refactor I think this is straightforward since ApplySettings seem intact.

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 2]
```diff
@@ -2116,6 +2117,33 @@ internal struct AvailableDevice
         internal IInputRuntime m_Runtime;
         internal InputMetrics m_Metrics;
         internal InputSettings m_Settings;
+
+        // Extract as booleans (from m_Settings) because feature check is in the hot path
```
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs:2121]: nice!

**Review by ritamerkl — APPROVED** [review]
Good change!

**Review by stefanunity — APPROVED** [review]
